File: R/ROptim.R
ROptim <- function(
    S,
    R,
    Rinv,
    lambda,
    tol = 1e-4,
    max_inner_iter = 10,
    max_outer_iter = 100) {
  if (is.null(Rinv)) {
    Rinv <- solve(R)
  }

  p <- dim(R)[1]
  lambda_matrix <- matrix(lambda, p, p); diag(lambda_matrix) <- 0
  ans <- ROptim_to_fortran(
    S,
    rho = lambda_matrix, thr = tol, maxIt = max_outer_iter,
    maxItLasso = max_inner_iter, start = "warm", Rinv_init = Rinv, R_init = R
  )

  if (ans$niter == max_outer_iter + 1) {
    rlang::warn(paste0("Optimization of R matrix reached the max number of iterations (", max_outer_iter, "). Consider increasing the `max_iter_R_outer` parameter in `pcglassoFast()` function."))
  }

  smallest_eigen_value <- eigen(ans$wi, TRUE, TRUE)$values[p]
  if (smallest_eigen_value < 0) {
    rlang::warn("Optimization of R matrix resulted in a matrix that is not positive definite. We highly encourage to increse the `max_iter_R_outer` parameter in `pcglassoFast()` function.")

    desired_smallest_eigen_value <- 0.01
    x <- (1-desired_smallest_eigen_value) / (1-smallest_eigen_value)
    ans$wi <- x*ans$wi + diag(1-x, p)
    ans$w <- solve(ans$wi)
  }

  list(
    R           = ans$wi,
    Rinv        = ans$w,
    outer.count = ans$niter,
    loglik      = NA
  )
}


# GLASSO algorithm of Friedman et al. 2008 with FORTRAN implementation of Sustik and Calderhead 2012.
# Ported to R by J. Clavel <julien.clavel@hotmail.fr> / <clavel@biologie.ens.fr> - 2017.

#' Fast optimization for the corelation matrix
#'
#' Part of the [pcglassoFast()] function.
#' This implementation is a small modification of the `glassoFast` function from `glassoFast` package.
#'
#' @param S Covariance matrix (a p by p symmetric matrix).
#' @param rho Regularization parameter (a non-negative value or a p by p matrix).
#' @param thr Threshold for convergence. Default is 1e-4.
#' @param maxIt Maximum number of iterations. Default is 10,000.
#' @param start Type of start: \code{"cold"} or \code{"warm"}.
#' @param Rinv_init Optional starting values for the estimated covariance matrix (p x p). Used only for warm starts.
#' @param R_init Optional starting values for the inverse covariance matrix (p x p). Used only for warm starts.
#' @param trace Logical. If \code{TRUE}, prints iteration info.
#'
#' @noRd
#'
#' @return A list with the following components:
#' \describe{
#'   \item{w}{Estimated covariance matrix}
#'   \item{wi}{Estimated inverse covariance matrix}
#'   \item{errflag}{Memory allocation error flag: 0 = no error}
#'   \item{niter}{Number of iterations}
#' }
#'
#' @keywords glasso covariance matrix regularization penalized likelihood
ROptim_to_fortran <- function(S, rho, thr = 1.0e-4, maxIt = 1e4, maxItLasso = 500, start = c("cold", "warm"), Rinv_init = NULL, R_init = NULL, trace = FALSE) {
  n <- nrow(S) # dimension of S
  if (is.matrix(rho)) {
    if (length(rho) != n * n) stop("The input matrix for \"rho\" must be of size ", n, " by ", n)
    L <- rho # matrix of regularization parameters
  } else {
    L <- matrix(rho, n, n) # matrix of regularization parameters
  }

  # cold or warm start
  switch(match.arg(start),
    cold = {
      is_flag <- 0
      W <- X <- matrix(0, n, n)
    },
    warm = {
      is_flag <- 1
      if (is.null(Rinv_init) || is.null(R_init)) {
        stop("Warm start specified: Rinv_init and R_init must be non-null")
      }
      W <- Rinv_init
      X <- R_init
    }
  )

  Wd <- WXj <- numeric(n)

  msg <- 1 * trace
  info <- 0
  mode(n) <- "integer"
  mode(S) <- "double"
  mode(L) <- "double"
  mode(thr) <- "double"
  mode(maxIt) <- "integer"
  mode(maxItLasso) <- "integer"
  mode(msg) <- "integer"
  mode(is_flag) <- "integer"
  mode(X) <- "double"
  mode(W) <- "double"
  mode(info) <- "integer"

  LASSO <- .Fortran(
    "roptim",
    n,
    S,
    L,
    thr,
    maxIt,
    maxItLasso,
    msg,
    is_flag,
    X,
    W,
    Wd,
    WXj,
    info
  )

  results <- list(w = LASSO[[10]], wi = LASSO[[9]], errflag = LASSO[[13]], niter = LASSO[[5]])
  return(results)
}

File: src/roptim.f90
      subroutine roptim(n,S,L,thr,maxIt,maxItLasso,msg,warm,X,W,Wd,WXj,info)
!
!     .. Scalar Arguments ..
      implicit double precision(a-h, o-z)
      integer n, warm, msg, maxIt, maxItLasso
!     ..
!     .. Array Arguments ..
      double precision S(n,n), L(n,n), X(n,n), W(n,n), Wd(n), WXj(n)
!     ..
!
!  Purpose
!  =======
!
!  TODO(Our comment)
!
!  TODO(Ensure X is positive-definite when maxIt is reached)
!
!  This subroutine computes the L1 regularized covariance matrix estimate
!  using the algorithm described in the paper:
!    J. Friedman, T. Hastie, R. Tibshirani:
!    Sparse inverse covariance estimation with the graphical lasso
!    Biostatistics, 9(3):432-441, July 2008.
!  This implementation is documented in the technical report:
!    M. A. Sustik B. Calderhead:
!    GLASSOFAST: An efficient GLASSO implementation
!    Technical Report TR-12-29, University of Texas at Austin
!
!  Arguments
!  =========
!
!  n      (input) integer
!         The dimension of the input matrix s
!
!  S      (input) double precision array, dimension n x n
!         The empirical covariance matrix
!
!  L      (input) double precision array, dimension n x n
!         Regularization matrix (symmetric);
!         We only tested this with the constand off-diagonal
!         and 0 diagonal
!
!  thr    (input) double precision
!         Convergence threshold
!
!  maxIt  (input/output) integer
!         Maximum number of whole matrix sweeps on input, actual number
!         of sweeps on output
!
!  maxItLasso (input) integer
!         Maximum number of inner LASSO passes per column
!
!  msg    (input) integer
!         Controls amount of messages printed
!
!  warm   (input) integer
!         flag indicating cold versus warm start, see also X, W
!
!  X      (input/output) double precision array, dimension n x n
!         Inverse covariance matrix estimate
!
!  W      (input/output) double precision array, dimension n x n
!         Covariance matrix estimate
!
!  Wd     (input) double precision array, dimension n
!         Internally used workspace
!
!  WXj    (input) double precision array, dimension n
!         Internally used workspace
!
!  info   (output) integer
!         Indicates errors
!
!  Wd, WDXj are not allocatable arrays, because gfortran/gdb has trouble
!  displaying those.

integer iter, innerIter
double precision EPS
parameter (EPS = 1.1e-16)
info = 0
shr = sum(abs(S))
do i = 1,n
   shr = shr - abs(S(i, i))
enddo
if (shr .eq. 0.0) then
!  S is diagonal.
   W = 0.0
   do i = 1,n
      W(i,i) = 1.0
   enddo
   X = 0.0
   do i = 1,n
      X(i,i) = 1.0
   enddo
   return
endif
shr = thr*shr/(n-1)
thrLasso = shr/n
if (thrLasso .lt. 2*EPS) then
   thrLasso = 2*EPS
end if
if (warm .eq. 0) then
   X = 0.0
   W = 0.0 ! TODO(Mayby different W would be better?)
   do i = 1,n
      W(i,i) = 1.0
   enddo
else
   do i = 1,n
      X(1:n,i) = -X(1:n,i)/X(i,i)
      X(i,i) = 0
   end do
end if
do iter = 1,maxIt
   if (msg .ne. 0)  call intpr('iter:',-1,iter,1)
   dw = 0.0
   do j = 1,n
      WXj(1:n) = 0.0
!     We exploit sparsity of X when computing column j of W*X*D:
      do i = 1,n
         if (X(i,j) .ne. 0.0) then
            WXj = WXj + W(:,i)*X(i,j)
         endif
      enddo
      innerIter = 0
      do
         innerIter = innerIter + 1
         dlx = 0.0
         do i = 1,n
            if (i .ne. j) then
               a = S(i,j) - WXj(i) + W(i,i)*X(i,j)
               b = abs(a) - L(i,j)
               if (b .gt. 0.0) then
                  c = sign(b, a)/W(i,i)
               else
                  c = 0.0
               endif
               delta = c - X(i,j)
               if (delta .ne. 0.0) then
                  X(i,j) = c
                  WXj(1:n) = WXj(1:n) + W(:,i)*delta
                  dlx = max(dlx, abs(delta))
               endif
            endif
         enddo
         if (dlx .lt. thrLasso) then
            exit
         endif
         if (innerIter .ge. maxItLasso) then
            exit
         endif
      enddo
      WXj(j) = W(j,j)
      dw = max(dw, sum(abs(WXj(1:n) - W(:,j))))
      W(:,j) = WXj(1:n)
      W(j,:) = WXj(1:n)
      dw = max(dw, abs(W(j,j) - (1 + sum(X(:,j)*W(:,j)))))
      W(j,j) = 1 + sum(X(:,j)*W(:,j))
   enddo
   if (dw .le. shr) then
      exit
   endif
enddo
do i = 1,n
   X(1:n,i) = -X(1:n,i)
   X(i,i) = 1
enddo
do i = 1,n-1
   X(i+1:n,i) = (X(i+1:n,i) + X(i,i+1:n))/2;
   X(i,i+1:n) = X(i+1:n,i)
enddo
maxIt = iter
return
end subroutine roptim
